File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode p1 = list1;
        ListNode p2 = list2;
        ListNode dummy = new ListNode();
        ListNode curr = dummy;
        while(p1 != null && p2 != null){
            if(p1.val <= p2.val){
                ListNode temp = p1.next;
                curr.next = p1;
                p1 = temp;
            }else{
                ListNode temp = p2.next;
                curr.next = p2;
                p2 = temp;
            }
            curr = curr.next;
        }
        if(p1 != null){
            curr.next = p1;
        }else{
            curr.next = p2;
        }
        return dummy.next;
    }
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        ptr1, ptr2 = list1, list2

        dummy = ListNode(0)
        prev = dummy

        while ptr1 and ptr2:
            if ptr1.val < ptr2.val:
                prev.next = ptr1
                ptr1 = ptr1.next
            else:
                prev.next = ptr2
                ptr2 = ptr2.next
            prev = prev.next

        if ptr1:
            prev.next = ptr1
        else:
            prev.next = ptr2

        return dummy.next